pkg/clusteragent/languagedetection/patcher.go

```diff
@@ -137,105 +149,151 @@ func (lp *languagePatcher) run() {
 			if !ok {
 				return
 			}
-			lp.handleEvent(eventBundle)
+			eventBundle.Acknowledge()
+			lp.handleEvents(eventBundle.Events)
 		}
 	}
 }
 
-func (lp *languagePatcher) generateAnnotationsPatch(currentLangs, newLangs langUtil.ContainersLanguages) map[string]interface{} {
-	currentAnnotations := currentLangs.ToAnnotations()
-	targetAnnotations := newLangs.ToAnnotations()
-
-	annotationsPatch := make(map[string]interface{})
+// handleEvents handles events from workloadmeta
+func (lp *languagePatcher) handleEvents(events []workloadmeta.Event) {
+	lp.logger.Tracef("Processing %d events", len(events))
 
-	// All newly detected languages should be included
-	for key, val := range targetAnnotations {
-		annotationsPatch[key] = val
-	}
-
-	// Languages that are no more detected should be removed
-	for key := range currentAnnotations {
-		_, found := annotationsPatch[key]
+	for _, event := range events {
+		owner, err := lp.extractOwnerFromEvent(event)
 
-		if !found {
-			annotationsPatch[key] = nil
+		if err != nil {
+			lp.logger.Errorf("failed to handle event: %v", err)
 		}
+
+		lp.queue.Add(*owner)
 	}
+}
 
-	return annotationsPatch
+func (lp *languagePatcher) extractOwnerFromEvent(event workloadmeta.Event) (*langUtil.NamespacedOwnerReference, error) {
+	entityKind := event.Entity.GetID().Kind
+
+	switch entityKind {
+	case workloadmeta.KindKubernetesDeployment:
+		// Case of deployment
+		deploymentID := event.Entity.(*workloadmeta.KubernetesDeployment).ID
+
+		// extract deployment name and namespace from entity id
+		deploymentIds := strings.Split(deploymentID, "/")
+		namespace := deploymentIds[0]
+		deploymentName := deploymentIds[1]
+
+		// construct namespaced owner reference
+		owner := langUtil.NewNamespacedOwnerReference(
+			"apps/v1",
+			langUtil.KindDeployment,
+			deploymentName,
+			namespace,
+		)
+
+		return &owner, nil
+	default:
+		return nil, fmt.Errorf("Unsupported entity kind: %v", entityKind)
+	}
 }
 
-// handleEvent handles events from workloadmeta
-func (lp *languagePatcher) handleEvent(eventBundle workloadmeta.EventBundle) {
-	eventBundle.Acknowledge()
-	lp.logger.Tracef("Processing %d events", len(eventBundle.Events))
+func (lp *languagePatcher) startProcessingPatchingRequests(ctx context.Context) {
+	go func() {
+		<-ctx.Done()
+		lp.queue.ShutDown()
+	}()
+	go func() {
+		for {
+			obj, shutdown := lp.queue.Get()
+			if shutdown {
+				break
+			}
+
+			owner, ok := obj.(langUtil.NamespacedOwnerReference)
+			if !ok {
+				// The item in the queue was not of the expected type. This should not happen.
```

Comment:
IMO we should have an error log or even a panic here as we are never supposed to execute that and we have full control over what is enqueued.

Reply:
I agree